### ISSUE TITLE
Design proposal: integrate SlaClip's private Slack Indicator and threshold controller

### ISSUE THREAD

**Issue by ZsyRock** [google-deepmind/jax_privacy#317]
Following maintainer guidance, I am opening this issue for design discussion before starting an implementation.

### 1. Proposal

I would like to contribute a paper-faithful implementation of [SlaClip: Gradient Norm Slacks can be Indicator for Adaptive Clipping in DP-SGD](https://icml.cc/virtual/2026/poster/66390) (ICML 2026).

SlaClip consists of two composable components:

1. A **private Slack Indicator mechanism**, which is the main technical contribution of SlaClip.
   For each per-example gradient `g_i`, clipping threshold `C_t`, and slack dimension `K`, it constructs `g_i^+ = (clip_C_t(g_i), s_i)`, where `s_i` encodes the unused norm margin in `K` bounded coordinates.
   The paper proves `||g_i^+||_2 <= C_t`, so its fixed-denominator average has the same `C_t / B` add/remove sensitivity as the ordinary clipped-gradient average.
   The gradient and Slack Indicator can therefore be released as coordinates of one Gaussian mechanism under the usual DP-SGD accounting, yielding a noisy, binned estimate of the gradient-norm CDF on `[0, C_t]` while leaving the ordinary DP gradient coordinates unchanged in distribution.
2. A **threshold controller** that consumes the noised Slack Indicator and applies the published update for `C_{t+1}` as post-processing.

The private Slack Indicator can be used independently, while composing both components gives the published SlaClip method.

Raw per-example norms, per-example slack, the exact empirical CDF, and the unnoised slack aggregate remain private intermediates and would not be returned or logged.

### 2. Proposed JAX Privacy integration

I propose preserving the same two component boundaries in JAX Privacy.

For the private Slack Indicator, the gradient and slack can remain separate PyTree leaves for computation and aggregation while being jointly noised through the existing DP query abstractions, so no flat `d + K` tensor is required.

For the controller, `C_t` would be maintained in immutable JAX state and updated only after the current noised Slack Indicator has been released.

My initial preference is a self-contained experimental implementation limited to scalar global clipping, explicit static `K`, and i.i.d. Gaussian noise.

Tests would verify the joint norm bound, the ordinary gradient path, the published controller, JAX transformations, and unchanged privacy accounting.

If maintainers prefer a reusable core hook or integration with the existing trainer, I would be happy to adjust the design accordingly.

### 3. Questions for maintainers

1. Does separating SlaClip into a private Slack Indicator mechanism and a threshold controller seem like the right design boundary for JAX Privacy?
2. For the initial contribution, would you prefer a self-contained implementation under `jax_privacy.experimental`, or integration with the existing core clipping and training abstractions?

I would be happy to adapt the design based on maintainer feedback and would like to implement the agreed approach.

**Comment by ryan112358** [comment]
Can you clarify one thing?

> noisy, binned estimate of the gradient-norm CDF on [0, C_t] while leaving the ordinary DP gradient coordinates unchanged in distribution.

Usually under DP-SGD, you compute the sum-of-clipped gradients, so I would expect with your approach you'd get a sum-of-grad-norm-slacks or similar.  How are you able to get the full CDF?

**Comment by ZsyRock** [comment]
> Can you clarify one thing?
> 
> > noisy, binned estimate of the gradient-norm CDF on [0, C_t] while leaving the ordinary DP gradient coordinates unchanged in distribution.
> 
> Usually under DP-SGD, you compute the sum-of-clipped gradients, so I would expect with your approach you'd get a sum-of-grad-norm-slacks or similar. How are you able to get the full CDF?

You are right that the aggregation produces sums of slack values. The key is that it produces `K` coordinate wise sums rather than one scalar sum.

For each per-example gradient, SlaClip scales its unused norm slack by `sqrt(K)` and encodes it into `K` extra coordinates. One way to visualize this is as pouring water into `K` bins in order, where each bin has capacity `lambda = C_t / sqrt(K)`: one bin is filled before moving to the next, until the encoded slack is exhausted. This produces a staircase shaped `K`-dimensional encoding, while the paper proves that the resulting extended gradient still has L2 norm at most `C_t`.

The fill level of each coordinate corresponds to a different gradient-norm interval, so its normalized batch sum gives the empirical CDF averaged over that interval. DP-SGD aggregates all `d + K` coordinates and jointly noises them with one isotropic Gaussian mechanism. The first `d` coordinates therefore have the same distribution as ordinary DP-SGD, while the last `K` coordinates provide a noisy, `K`-bin CDF estimate over `[0, C_t]`.

So this does not recover the exact or continuous full CDF; it provides `K` bin-averaged CDF values. The formal construction, norm-bound proof, and choice of `K` are given in the paper.

**Comment by ryan112358** [comment]
Do I understand correctly that if clip norm is 1.0 and you have K=10 then each extra coordinate would correspond to [0, 0.1), [0.1, 0.2), [0.2, 0.3), ... 

Can you show me a worked-out example where the slack is 0.25, how does the mass get distributed concretely?

To answer your questions: 
1. Not sure I understand
2. Strong preference for integrating with existing abstractions where feasible.  If JAX Privacy lacks flexibility to do this cleanly, that is something we'd like to fix.

Assuming my understanding is correct, is the "pre_clipping_transform" defined on clipped_grad and used on optimization.py sufficiently flexible to represent your mechanism (possibly combined with per_leaf_clipping).  The other possibility would be to add a "slack_k" optional arg to clipped_grad that defaults to None.  Maybe you have some other design ideas

**Comment by ZsyRock** [comment]
> Do I understand correctly that if clip norm is 1.0 and you have K=10 then each extra coordinate would correspond to [0, 0.1), [0.1, 0.2), [0.2, 0.3), ...
> 
> Can you show me a worked-out example where the slack is 0.25, how does the mass get distributed concretely?
> 
> To answer your questions:
> 
> 1. Not sure I understand
> 2. Strong preference for integrating with existing abstractions where feasible.  If JAX Privacy lacks flexibility to do this cleanly, that is something we'd like to fix.
> 
> Assuming my understanding is correct, is the "pre_clipping_transform" defined on clipped_grad and used on optimization.py sufficiently flexible to represent your mechanism (possibly combined with per_leaf_clipping). The other possibility would be to add a "slack_k" optional arg to clipped_grad that defaults to None. Maybe you have some other design ideas

Thanks. The intended design has two components:

component 1. The Slack Indicator extends each per-example gradient to

    [Clip_C(g_i); s_i]

and releases the gradient and slack coordinates together through the same Gaussian mechanism. Since the extended vector still has L2 norm at most C, this gives a noisy, binned CDF estimate with no additional private query or privacy cost beyond the main DP-SGD release. This component can be reused independently.

component 2. The SlaClip controller consumes only the noisy CDF and updates C_{t+1} as post-processing. Other methods could instead use the same CDF of component 1 to update C, the learning rate, or another parameter.

For the toy example you mentioned, C = 1 and K = 10, the norm-bin width is 0.1. The paper stores the coordinates in reverse order:

    coordinate 1: [0.9, 1.0]
    coordinate 2: [0.8, 0.9]
    ...
    coordinate 10: [0, 0.1].

The coordinate capacity is

    lambda = C / sqrt(K) = 1 / sqrt(10) ≈ 0.316227.

This is the largest capacity that preserves the extended-vector norm bound and, for fixed K, minimizes the normalized noise.

For slack 0.25,

    sqrt(K) * slack
      = sqrt(10) * 0.25
      ≈ 0.790569
      = 2 * lambda + 0.158114,

so the extended vector will be formed as:

    s_i = [0.316227, 0.316227, 0.158114, 0, ..., 0].

If ||g_i|| = 0.75, then

    ||[Clip_C(g_i); s_i]||_2 ≈ 0.8874 <= 1, still smaller than C_t =1.

Every example receives its own prefix-filled slack vector. For example, samples with different gradient norms may produce

    s_1 = [0.316227, 0.316227, 0.158114, 0, ...],
    s_2 = [0.316227, 0.1, 0, ...],
    s_3 = [0.316227, 0.316227, 0.316227, 0.2, 0, ...].

Across a batch, these slack vectors are released jointly with the gradients. Using average notation,

    tilde_s
      = (1/B) * sum_i s_i
        + N(0, (sigma*C/B)^2 I_K),

and the Slack Indicator is

    s_hat = tilde_s / lambda, is the noisy binned CDF estimate. For example,

    s_hat = [0.60, 0.55, 0.48, ..., 0.10]

means that the CDF near C is approximately 0.60. With a sufficiently fine bin, this corresponds to approximately 60% of examples being below C, or approximately 40% being clipped.

For the JAX Privacy integration, I think pre_clipping_transform is sufficient for an initial implementation. It could return

    gradient = Clip_C(g_i) / C,
    slack = s_i / C,

as one PyTree with joint norm at most one. The existing code could then clip, aggregate, and noise both leaves jointly. The augmented optimizer would use the noisy gradient for optimization and optionally pass the noisy CDF to a controller.

I would initially support scalar global clipping only. If this prototype reveals that the noisy auxiliary output cannot be exposed cleanly, we could then add a small generic hook rather than a SlaClip-specific slack_k argument.

**Comment by ryan112358** [comment]
This plan sounds good to me, and thank you for the detailed explanation!  The only drawback of pre_clipping_transform is that the clipping operation might happen twice, and I'm not sure if XLA will optimize it away or not.  If this turns out to be a major overhead, we can consider upstreaming directly to clipping.py.  Sounds good on supporting global clipping for now.  

One question worth thinking about is how might this compose with other pre_clipping_transforms?  For now the primary one we implement is ScaleThenPrivatize, but there's many others: GeoClip, DelayedPreconditioners, etc. It would be nice if you didn't have to pick SlaClip OR one of these, but if you could use both at the same time.  

Feel free to open up a PR, I think I have enough information and agree your plan is solid.

**Comment by ZsyRock** [comment]
> This plan sounds good to me, and thank you for the detailed explanation! The only drawback of pre_clipping_transform is that the clipping operation might happen twice, and I'm not sure if XLA will optimize it away or not. If this turns out to be a major overhead, we can consider upstreaming directly to clipping.py. Sounds good on supporting global clipping for now.
> 
> One question worth thinking about is how might this compose with other pre_clipping_transforms? For now the primary one we implement is ScaleThenPrivatize, but there's many others: GeoClip, DelayedPreconditioners, etc. It would be nice if you didn't have to pick SlaClip OR one of these, but if you could use both at the same time.
> 
> Feel free to open up a PR, I think I have enough information and agree your plan is solid.

Thanks. I implemented and tested the pre_clipping_transform version we discussed. It correctly implements the joint gradient-and-slack release and the optional controller, but it still has a significant performance limitation.

On an RTX 4060 with two synthetic workloads, this version adds about 33.1% and 51.3% overhead. The per-example gradient is differentiated only once. The overhead occurs because SlaClip computes the norm and constructs a bounded gradient-and-slack query inside pre_clipping_transform, after which the generic clipping path performs another full norm, scaling, and PyTree traversal.

I also isolated the cost of the additional slack coordinates. In the reduction-heavy benchmark, increasing K from 1 to 100 changed the median step time only from 2.1216 ms to 2.1247 ms, approximately 0.15%, with no material memory difference. In this workload, the K-dimensional slack computation is therefore not the main source of overhead. This is expected because K <= 100 is very small compared with approximately 1.05 million gradient coordinates, whereas the repeated clipping traversal scales with the full model gradient.

A controlled diagnostic that bypasses only the second full clipping traversal reduces the remaining overhead to about 0.3% and 11.8%, respectively. This diagnostic is not yet production-ready, but it confirms that the repeated full clipping traversal is the main optimization target.

My proposed production version would add a small generic hook, likely in clipping.py, that can safely reuse the already computed norm/bound. It would preserve NaN/Inf handling, padding, composition with other transforms, and the original code path when the hook is unused. The existing two-component SlaClip API and joint DP release would remain unchanged.

Would you prefer that I implement this optimized integration and then open one draft PR, or first open the current pre_clipping_transform prototype as a draft for API review?

**Comment by ryan112358** [comment]
Can you provide more details about the experimental setup (model, batch size, etc.).  And did you apply @jax.jit to your function?

Given the non-trivial overhead, adding something in clipping.py could make sense, so I'd like to better understand what your generic hool would look like.  One idea is to add an optional kwarg to clipped_grad and update the return namedtuple to include a "slack" output or similar.  Since it's free, we can maybe even make it the default (or just always on, no kwarg)

**Comment by ZsyRock** [comment]
> Can you provide more details about the experimental setup (model, batch size, etc.).  And did you apply @jax.jit to your function?
> 
> Given the non-trivial overhead, adding something in clipping.py could make sense, so I'd like to better understand what your generic hool would look like.  One idea is to add an optional kwarg to clipped_grad and update the return namedtuple to include a "slack" output or similar.  Since it's free, we can maybe even make it the default (or just always on, no kwarg)

Yes. These are steady state JIT-compiled timings. `DPTrainer.train_step` is decorated with `@jax.jit`, and each process explicitly called `.lower(...).compile()`. Compilation was excluded, followed by 50 warm up steps and synchronization before and after timing.

Baseline and SlaClip used the same float32 setup with batch size 32 on an RTX 4060. The workloads were a 1,048,577 parameter logistic regression and a 1,050,625 parameter 512->1024->512->1 MLP. The SlaClip runs used a slack dimension of K=10, non-private per-example auxiliary outputs were disabled, and the reported overheads are based on seven paired fresh process runs per workload. These were fixed batch `train_step` benchmarks rather than end-to-end training measurements.

I agree that an optional `slack_k` argument to `clipped_grad` is likely the right public API, initially for scalar global clipping. The existing clipping path already obtains `(clipped_gradient, norm)`, so it can reuse that norm to construct slack without a second full gradient clipping traversal. Gradient and slack would form a bounded primary output such as `ClippedGradOutput(gradient, slack)` and be aggregated and noised jointly. Slack should not be placed in the existing `AuxiliaryOutput`, because that output is per-example and is not part of the bounded/noised query.

I would keep `slack_k=None` by default: slack is free in privacy accounting, but not computationally or structurally free. This preserves the existing output, execution, and checkpoint structure when SlaClip is disabled. When enabled, slack construction would occur after the existing `pre_clipping_transform`, preserving composition with it. The implementation would include the trainer plumbing needed to initialize and noise the joint query and route only the noisy slack to the optional controller.

If this matches your intended API, I will implement it, rerun the correctness and performance tests, and open a draft PR.

**Comment by ryan112358** [comment]
This plan makes sense to me, happy to review a PR that does this!

**Comment by ZsyRock** [comment]
Thanks — I’ve opened draft PR #343 with the optimized single-pass `slack_k` integration: https://github.com/google-deepmind/jax_privacy/pull/343.

It reuses the norm already computed by `clipped_grad`, jointly noises the gradient and slack output, composes after `pre_clipping_transform`, and preserves the existing path when disabled. In the final compile-excluded GPU benchmarks, CDF-only overhead was 0.09%–3.32% and adaptive overhead was 0.15%–3.30%. The API, privacy argument, tests, and limitations are documented in the PR.

**Comment by ryan112358** [comment]
@ZsyRock your first PR was just merged!  We are hoping to improve our contributor experience in the future, thanks for your patience in working with us on this.  Looking forward to seeing the next PR!

**Comment by ZsyRock** [comment]
Thanks Ryan! I really appreciate your guidance and patience throughout the review. I’m glad the privacy-critical Slack Indicator primitive is now merged. Since threshold adaptation is post-processing and may admit different controller designs, I plan to keep the core interface general and leave controller integration to downstream research for now. I’d be happy to revisit it later if there is a clear use case for JAX Privacy.

**Comment by ryan112358** [comment]
Sounds good, makes sense!
